core/amber/src/main/scala/edu/uci/ics/texera/web/resource/dashboard/admin/execution/AdminExecutionResource.scala

```diff
@@ -45,49 +47,137 @@ object AdminExecutionResource {
-    val workflowEntries = context
+  def listWorkflows(
+      @Auth current_user: SessionUser,
+      @PathParam("pageSize") page_size: Int = 20,
+      @PathParam("pageIndex") page_index: Int = 0,
+      @PathParam("sortField") sortField: String = "end_time",
+      @PathParam("sortDirection") sortDirection: String = "desc",
+      @QueryParam("filter") filter: java.util.List[String]
+  ): List[dashboardExecution] = {
+    val filter_status = filter.asScala.map(mapToStatus).toSeq.filter(_ != -1).asJava
+
+    // Retrieve the latest execution id for each workflow
+    val latestExecutionId = context
+      .select(
+        WORKFLOW_VERSION.WID,
+        DSL.max(WORKFLOW_EXECUTIONS.EID).as("max_eid")
+      )
+      .from(WORKFLOW_EXECUTIONS)
+      .join(WORKFLOW_VERSION)
+      .on(WORKFLOW_VERSION.VID.eq(WORKFLOW_EXECUTIONS.VID))
+      .groupBy(WORKFLOW_VERSION.WID)
+      .asTable("latest_execution_id")
+
+    // Base query that retrieves latest execution info for each workflow without sorting and filtering.
+    // Only retrieving executions in current page according to pageSize and pageIndex parameters.
+    val executions_base_query = context
       .select(
         WORKFLOW_EXECUTIONS.UID,
         USER.NAME,
         WORKFLOW_VERSION.WID,
         WORKFLOW.NAME,
         WORKFLOW_EXECUTIONS.EID,
-        WORKFLOW_EXECUTIONS.VID,
         WORKFLOW_EXECUTIONS.STARTING_TIME,
         WORKFLOW_EXECUTIONS.LAST_UPDATE_TIME,
         WORKFLOW_EXECUTIONS.STATUS,
         WORKFLOW_EXECUTIONS.NAME
       )
       .from(WORKFLOW_EXECUTIONS)
-      .leftJoin(WORKFLOW_VERSION)
+      .join(WORKFLOW_VERSION)
       .on(WORKFLOW_EXECUTIONS.VID.eq(WORKFLOW_VERSION.VID))
-      .leftJoin(USER)
+      .join(USER)
       .on(WORKFLOW_EXECUTIONS.UID.eq(USER.UID))
-      .leftJoin(WORKFLOW)
+      .join(WORKFLOW)
       .on(WORKFLOW.WID.eq(WORKFLOW_VERSION.WID))
-      .fetch()
+      .join(latestExecutionId)
+      .on(
+        DSL.and(
+          WORKFLOW_EXECUTIONS.EID
+            .eq(latestExecutionId.field("max_eid").asInstanceOf[Field[UInteger]]),
+          WORKFLOW_VERSION.WID.eq(latestExecutionId.field(WORKFLOW_VERSION.WID))
+        )
+      )
+
+    // Apply filter if the status are not empty.
+    val executions_apply_filter = if (!filter_status.isEmpty) {
+      executions_base_query.where(WORKFLOW_EXECUTIONS.STATUS.in(filter_status))
+    } else {
+      executions_base_query
+    }
+
+    // Apply sorting if user specified.
+    var executions_apply_order =
+      executions_apply_filter.limit(page_size).offset(page_index * page_size)
+    if (sortField != "NO_SORTING") {
+      val orderByField = sortFieldMapping.getOrElse(sortField, WORKFLOW.NAME)
+      val order = if (sortDirection == "desc") orderByField.desc() else orderByField.asc()
+      executions_apply_order =
+        executions_apply_filter.orderBy(order).limit(page_size).offset(page_index * page_size)
```

Comment:
This code segment can be merged into a single line.
`.limit(page_size).offset(page_index * page_size)`

Reply:
Wouldn't this make the code too complex in one line?

Reply:
addressed